### PR TITLE
python3Packages.PyLaTeX: init at 1.4.2

### PR DESCRIPTION
**Pull request by MayNiklas**
###### Description of changes

The documentation describes this library the best: "PyLaTeX is a Python library for creating and compiling LaTeX files. The goal of this library is to be an easy, but extensible interface between Python and LaTeX.".

https://jeltef.github.io/PyLaTeX/current/index.html
https://github.com/JelteF/PyLaTeX

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by MayNiklas** [comment]
Since the dependencies are used at runtime, I replaced `BuildInputs` with `propagatedBuildInputs`.

File: pkgs/development/python-modules/pylatex/default.nix
Comment (fabaff) [inline pkgs/development/python-modules/pylatex/default.nix:14]:
```suggestion
  pname = "pylatex";
```

File: pkgs/development/python-modules/pylatex/default.nix
Comment (fabaff) [inline pkgs/development/python-modules/pylatex/default.nix:15]: We should stick to supported Python releases for new packages..

```suggestion
  version = "1.3.4";
  format = "setuptools";

  disabled = pythonOlder "3.7";
```

File: pkgs/development/python-modules/pylatex/default.nix
Comment (fabaff) [inline pkgs/development/python-modules/pylatex/default.nix:27]: Please use `passthru.optional-dependencies` instead of flags.

**Review by fabaff — COMMENTED** [review]
Please add `pythonImportsCheck` and try to use `pytestCheckHook`, if possible, to run the tests.

**Comment by MayNiklas** [comment]
@fabaff could you maybe take another look at this?
I would love to get this package into the next release.

File: pkgs/top-level/python-packages.nix
Comment (necrophcodr) [inline pkgs/top-level/python-packages.nix:8039]: It might make sense to ensure these are properly ordered alphabetically as well.

**Review by necrophcodr — CHANGES_REQUESTED** [review]

**Comment by necrophcodr** [comment]
@MayNiklas it looks pretty decent. I don't see anything that doesn't fit the contributors guidelines, so if this is still something you want, I think it makes sense to have it polished up and ready for merging.

File: pkgs/top-level/python-packages.nix
Comment (MayNiklas) [reply pkgs/top-level/python-packages.nix:8806]: You're right!
I rebased with the master branch and made sure the packages are ordered correctly!

**Comment by MayNiklas** [comment]
> @MayNiklas it looks pretty decent. I don't see anything that doesn't fit the contributors guidelines, so if this is still something you want, I think it makes sense to have it polished up and ready for merging.

That sounds great!
I rebased against the master branch and made sure, the packages around it are sorted correctly.

I still think, this package makes sense in NixPkgs!
-> I use it to create PDF files (tables & invoices) with python scripts
-> since it depends on Latex, having everything locked by a flake.nix is great!
-> even if there are no current commits in the repository, it works great (once you get behind it)!

**Review by necrophcodr — APPROVED** [review]
I don't see anything obvious missing, seems pretty good!

File: pkgs/development/python-modules/pylatex/default.nix
Comment (pbsds) [inline pkgs/development/python-modules/pylatex/default.nix:34]: Should these be part of the default runtime closure, or are they only needed at check time? If the latter, consider moving them to `checkInputs`

File: pkgs/development/python-modules/pylatex/default.nix
Comment (pbsds) [inline pkgs/development/python-modules/pylatex/default.nix:28]: Names should mach the corresponding extras [here](https://github.com/JelteF/PyLaTeX/blob/2050b14d8d7ed4fe788c769afec6816e2b703355/setup.py#L30-L37)

**Review by pbsds — CHANGES_REQUESTED** [review]
Good work, nearly there!

File: pkgs/development/python-modules/pylatex/default.nix
Comment (MayNiklas) [reply pkgs/development/python-modules/pylatex/default.nix:34]: They should be present at runtime.
A lot of features depend on those libraries. For example to translate matrices from numpy into LaTex.

File: pkgs/development/python-modules/pylatex/default.nix
Comment (MayNiklas) [reply pkgs/development/python-modules/pylatex/default.nix]: > Names should mach the corresponding extras [here](https://github.com/JelteF/PyLaTeX/blob/2050b14d8d7ed4fe788c769afec6816e2b703355/setup.py#L30-L37)

Should be fixed now 👍

File: pkgs/development/python-modules/pylatex/default.nix
Comment (pbsds) [reply pkgs/development/python-modules/pylatex/default.nix]: The please add a comment stating that you've deemed their inclusion common enough to be the default.

File: pkgs/development/python-modules/pylatex/default.nix
Comment (FRidh) [inline pkgs/development/python-modules/pylatex/default.nix:31]: Please avoid adding optional dependencies by default. Users can easily add packages to an environment, but removing packages is non-trivial.

Comment (MayNiklas) [reply pkgs/development/python-modules/pylatex/default.nix:31]: Great point!
I'm removing those dependencies so people using the library can just add what they need.

File: pkgs/development/python-modules/pylatex/default.nix
Comment (MayNiklas) [reply pkgs/development/python-modules/pylatex/default.nix]: I did :)

**Review by pbsds — APPROVED** [review]

**Comment by MayNiklas** [comment]
@FRidh could you maybe please take another look at this PR?
I implemented all the changes discussed.

Should be finally ready to merge.

File: pkgs/development/python-modules/pylatex/default.nix
Comment (natsukium) [inline pkgs/development/python-modules/pylatex/default.nix:11]:
```suggestion
  pyproject = true;
```
Could you please use `pyproject = true` instead?
https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md#buildpythonpackage-parameters-buildpythonpackage-parameters

While waiting for the merge, the recommended writing style has changed. Sorry for the extra work.

File: pkgs/development/python-modules/pylatex/default.nix
Comment (natsukium) [inline pkgs/development/python-modules/pylatex/default.nix:10]: FYI: 1.4.2 has been released. https://github.com/JelteF/PyLaTeX/tree/v1.4.2

It does not need to be worked on by this PR, but may be worthwhile as python 3.12 will be coming to mainstream nixpkgs python in the near future.

File: pkgs/development/python-modules/pylatex/default.nix
Comment (MayNiklas) [reply pkgs/development/python-modules/pylatex/default.nix]: Done!

File: pkgs/development/python-modules/pylatex/default.nix
Comment (MayNiklas) [reply pkgs/development/python-modules/pylatex/default.nix]: I changed that as well.

**Comment by natsukium** [comment]
@ofborg build python310Packages.pylatex python311Packages.pylatex

File: pkgs/development/python-modules/pylatex/default.nix
Comment (natsukium) [inline pkgs/development/python-modules/pylatex/default.nix:21]:
```suggestion

  nativeBuildInputs = [
    setuptools
  ];
```

File: pkgs/development/python-modules/pylatex/default.nix
Comment (pbsds) [inline pkgs/development/python-modules/pylatex/default.nix:33]:
```suggestion

  nativeCheckInputs = [
    pytestCheckHook
    matplotlib
    quantities
    (texlive.combine { inherit (texlive)
      scheme-small
      lastpage
      collection-fontsrecommended
    ;})
  ];

```

File: pkgs/development/python-modules/pylatex/default.nix
Comment (pbsds) [inline pkgs/development/python-modules/pylatex/default.nix:6]:
```suggestion
, ordered-set
, pytestCheckHook
, matplotlib
, quantities
, texlive
```

**Review by pbsds — APPROVED** [review]
Result of `nixpkgs-review pr 203093` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>python310Packages.pylatex</li>
    <li>python310Packages.pylatex.dist</li>
    <li>python311Packages.pylatex</li>
    <li>python311Packages.pylatex.dist</li>
  </ul>
</details>

```
[nix-shell:/dev/shm/nixpkgs-review/pr-203093]$ nix-shell -I nixpkgs=nixpkgs/ -p python3Packages.pylatex --pure --run python3
Python 3.11.6 (main, Oct  2 2023, 13:45:54) [GCC 12.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import pylatex
>>> pylatex.
pylatex.Alignat(              pylatex.HorizontalSpace(      pylatex.NewPage(              pylatex.TextBlock(            pylatex.errors
pylatex.Autoref(              pylatex.HugeText(             pylatex.NoEscape(             pylatex.TextColor(            pylatex.escape_latex(
pylatex.Axis(                 pylatex.Hyperref(             pylatex.Package(              pylatex.TikZ(                 pylatex.figure
pylatex.Center(               pylatex.Itemize(              pylatex.PageStyle(            pylatex.TikZCoordinate(       pylatex.frames
pylatex.Chapter(              pylatex.Label(                pylatex.Pageref(              pylatex.TikZDraw(             pylatex.headfoot
pylatex.ColumnType(           pylatex.LargeText(            pylatex.Plot(                 pylatex.TikZNode(             pylatex.labelref
pylatex.Command(              pylatex.LineBreak(            pylatex.Quantity(             pylatex.TikZNodeAnchor(       pylatex.lists
pylatex.Description(          pylatex.LongTable(            pylatex.Ref(                  pylatex.TikZOptions(          pylatex.math
pylatex.Document(             pylatex.LongTabu(             pylatex.Section(              pylatex.TikZPath(             pylatex.package
pylatex.Enumerate(            pylatex.LongTabularx(         pylatex.SmallText(            pylatex.TikZPathList(         pylatex.position
pylatex.Eqref(                pylatex.Marker(               pylatex.StandAloneGraphic(    pylatex.TikZScope(            pylatex.quantities
pylatex.FBox(                 pylatex.Math(                 pylatex.SubFigure(            pylatex.TikZUserPath(         pylatex.section
pylatex.Figure(               pylatex.Matrix(               pylatex.Subsection(           pylatex.UnsafeCommand(        pylatex.simple_page_number()
pylatex.FlushLeft(            pylatex.MdFramed(             pylatex.Subsubsection(        pylatex.VectorName(           pylatex.table
pylatex.FlushRight(           pylatex.MediumText(           pylatex.Table(                pylatex.VerticalSpace(        pylatex.tikz
pylatex.Foot(                 pylatex.MiniPage(             pylatex.TableRowSizeError(    pylatex.base_classes          pylatex.utils
pylatex.FootnoteText(         pylatex.MultiColumn(          pylatex.Tabu(                 pylatex.basic
pylatex.HFill(                pylatex.MultiRow(             pylatex.Tabular(              pylatex.config
pylatex.Head(                 pylatex.NewLine(              pylatex.Tabularx(             pylatex.document
```

Build results look fine, and the derivation seem looks fine for me.

I managed to make the tests run as well, please apply the suggestions.

**Comment by MayNiklas** [comment]
> I managed to make the tests run as well, please apply the suggestions.

@pbsds Thank you!!!

**Comment by natsukium** [comment]
@ofborg build python310Packages.pylatex　python311Packages.pylatex

**Review by natsukium — APPROVED** [review]
LGTM, thanks.
I'll merge it after ofborg seems fine.

**Pull request merged**
